lib/krb5/mk_cred.c

```diff
@@ -0,0 +1,323 @@
+               krb5_creds **ppcreds,
+               krb5_data *out_data,
+               krb5_replay_data *replay_data)
+{
+    krb5_error_code ret;
+    EncKrbCredPart enc_krb_cred_part;
+    KrbCredInfo *krb_cred_info;
+    krb5_crypto crypto;
+    KRB_CRED cred;
+    unsigned char *buf = NULL;
+    size_t ncreds, i;
+    size_t buf_size;
+    size_t len;
+
+    /*
+     * The ownership of 'buf' is re-assigned to a containing structure
+     * multiple times. We enforce an invariant, either buf is non-zero
+     * and we own it, or buf is zero and it is freed or some structure
+     * owns any storage previously allocated as 'buf'.
+     */
+#define CHOWN_BUF(x, buf) do { (x) = (buf); (buf) = 0; } while (0)
+#define DISOWN_BUF(buf) do { free(buf); (buf) = 0; } while (0)
+
+    for (ncreds = 0; ppcreds[ncreds]; ncreds++)
+        ;
+
+    memset (&cred, 0, sizeof(cred));
+    cred.pvno = 5;
+    cred.msg_type = krb_cred;
+    ALLOC_SEQ(&cred.tickets, ncreds);
+    if (cred.tickets.val == NULL) {
+        ret = krb5_enomem(context);
+        goto out;
+    }
+    memset (&enc_krb_cred_part, 0, sizeof(enc_krb_cred_part));
+    ALLOC_SEQ(&enc_krb_cred_part.ticket_info, ncreds);
+    if (enc_krb_cred_part.ticket_info.val == NULL) {
+        ret = krb5_enomem(context);
+        goto out;
+    }
+
+    for (i = 0; i < ncreds; i++) {
+        ret = decode_Ticket(ppcreds[i]->ticket.data,
+                            ppcreds[i]->ticket.length,
+                            &cred.tickets.val[i],
+                            &len);/* don't care about len */
+        if (ret)
+           goto out;
+
+        /* fill ticket_info.val[i] */
+        krb_cred_info = &enc_krb_cred_part.ticket_info.val[i];
+
+        CHECKED_COPY(copy_EncryptionKey,
+                     &krb_cred_info->key, &ppcreds[i]->session);
+        CHECKED_ALLOC_COPY_PPC2KCI(copy_Realm, prealm, client->realm);
+        CHECKED_ALLOC_COPY_PPC2KCI(copy_PrincipalName, pname, client->name);
+        CHECKED_ALLOC_ASSIGN_PPC2KCI(flags, flags.b);
+        CHECKED_ALLOC_ASSIGN_PPC2KCI(authtime, times.authtime);
+        CHECKED_ALLOC_ASSIGN_PPC2KCI(starttime, times.starttime);
+        CHECKED_ALLOC_ASSIGN_PPC2KCI(endtime, times.endtime);
+        CHECKED_ALLOC_ASSIGN_PPC2KCI(renew_till, times.renew_till);
+        CHECKED_ALLOC_COPY_PPC2KCI(copy_Realm, srealm, server->realm);
+        CHECKED_ALLOC_COPY_PPC2KCI(copy_PrincipalName, sname, server->name);
+        CHECKED_ALLOC_COPY_PPC2KCI(copy_HostAddresses, caddr, addresses);
+    }
+
+    if (auth_context->flags & KRB5_AUTH_CONTEXT_DO_TIME) {
+        krb5_timestamp sec;
+        int32_t usec;
+
+        krb5_us_timeofday (context, &sec, &usec);
+
+        CHECKED_ALLOC_ASSIGN(enc_krb_cred_part.timestamp, sec);
+        CHECKED_ALLOC_ASSIGN(enc_krb_cred_part.usec, usec);
+    } else {
+        enc_krb_cred_part.timestamp = NULL;
+        enc_krb_cred_part.usec = NULL;
+        /* XXX Er, shouldn't we set the seq nums?? */
```

Comment:
For GSS this won't matter.  For other protocols, it might, but we can fix it if we ever need it (we won't).